core/src/banking_stage/thread_aware_account_locks.rs

```diff
@@ -0,0 +1,692 @@
+        // that is not at the sequential limit.
+        self.schedulable_threads_with_read_only_handler(account, |thread_set, counts| {
+            let mut schedulable_threads = ThreadSet::any(self.num_threads);
+            for thread_id in thread_set.threads_iter() {
+                if counts[thread_id] == self.sequential_queue_limit {
+                    schedulable_threads.remove(thread_id);
+                }
+            }
+            schedulable_threads
+        })
+    }
+
+    /// Returns `ThreadSet` of schedulable threads for the given writable account.
+    /// If the account is not locked, then all threads are schedulable.
+    /// If read-locked on a single thread, then only that thread is schedulable.
+    /// If write-locked, then only that thread is schedulable.
+    /// In all other cases, no threads are schedulable.
+    /// The sequential limit is checked, and a thread will not be returned as schedulable
+    /// if the limit is reached.
+    fn write_schedulable_threads(&self, account: &Pubkey) -> ThreadSet {
+        // If the account is only read locked, then a write lock can only be taken
+        // if the read lock is held by a single thread, and the limit is not exceeded.
+        self.schedulable_threads_with_read_only_handler(account, |thread_set, counts| {
+            thread_set
+                .only_one_scheduled()
+                .filter(|thread_id| counts[*thread_id] < self.sequential_queue_limit)
+                .map_or_else(ThreadSet::none, ThreadSet::only)
+        })
+    }
+
+    /// Returns `ThreadSet` of schedulable threads, given the read-only lock handler.
+    /// Helper function, since the only difference between read and write schedulable threads
+    /// is in how the case where only read locks are held is handled.
+    /// If there are no locks, then all threads are schedulable.
+    /// If only write-locked, then only the thread holding the write lock is schedulable.
+    /// If a mix of locks, then only the write thread is schedulable.
+    /// The sequential limit is checked, and a thread will not be returned as schedulable
+    /// if the limit is reached.
+    fn schedulable_threads_with_read_only_handler(
+        &self,
+        account: &Pubkey,
+        read_only_handler: impl Fn(&ThreadSet, &[u32]) -> ThreadSet,
+    ) -> ThreadSet {
+        match (self.write_locks.get(account), self.read_locks.get(account)) {
+            (None, None) => ThreadSet::any(self.num_threads),
+            (None, Some((thread_set, counts))) => read_only_handler(thread_set, counts),
+            (Some((thread_id, count)), None) => {
+                if count == &self.sequential_queue_limit {
+                    ThreadSet::none()
+                } else {
+                    ThreadSet::only(*thread_id)
+                }
+            }
+            (Some((thread_id, count)), Some((thread_set, counts))) => {
+                debug_assert_eq!(Some(*thread_id), thread_set.only_one_scheduled());
```

Comment:
nits: why `debug_assert_eq!` is used instead of plain `assert_eq!`?

Reply:
`only_one_scheduled` is one of the more expensive operations on the thread_set (though probably cheap).

It should not be possible (using the public interfaces) to enter a situation where this occurs, so was just trying to leave it as a test-only check.

Not too opposed to making it assert_eq, as I didn't actually measure the affect. What do you think?

Reply:
converted to an assert: 4ad105b13ce68b1ed3c0c8cd8fce6e1758be1128